### PR TITLE
Improved docs for module.py + multiple cleanups in docs

### LINKED ISSUE #1177: 🐛[BUG]: UNet example not working with `model.save()` because of name clashes
### Version

1.2.0

### On which installation method(s) does this occur?

_No response_

## Problem

@nbren reported that the [UNet example](https://docs.nvidia.com/physicsnemo/latest/physicsnemo/api/models/modules.html#how-to-write-your-own-physicsnemo-model) in the documentation is not working with `model.save()` because of some clashes with the `UNet` class shipped in the physicsnemo model zoo.

It is unclear why the names are clashing. Is it because it is not possible to define a subclass of `physicsnemo.Module` whose `class.__name__` is already in the model registry?

## Solution

- If names clashes cannot be avoided, this should at least be documented, and the error message should be improved to be more informative on the reason of the clash.

### Minimum reproducible example

```shell

```

### Relevant log output

```shell

```

### Environment details

```shell

```

### PR DESCRIPTION
**Pull request by CharlelieLrt**
# PhysicsNeMo Pull Request

## Description

Closes #1177 #1178 #1176 #1169 .

## Checklist

- [ ] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: docs/api/models/modules.rst [hunk 1]
```diff
@@ -9,22 +9,34 @@ PhysicsNeMo Modules
...
 
-PhysicsNeMo contains its own Model class for constructing neural networks. This model class
+PhysicsNeMo contains its own Module class for constructing neural networks. This class
 is built on top of PyTorch's ``nn.Module`` and can be used interchangeably within the
-PyTorch ecosystem. Using PhysicsNeMo models allows you to leverage various features of
-PhysicsNeMo aimed at improving performance and ease of use. These features include, but are
-not limited to, model zoo, automatic mixed-precision, CUDA Graphs, and easy checkpointing.
-We discuss each of these features in the following sections.
+PyTorch ecosystem. Using PhysicsNeMo modules allows you to leverage various features
+aimed at improving performance and ease of use. These features include, but are
```
Comment (ktangsali) [inline docs/api/models/modules.rst]: Given that we are not emphasizing on Static Capture anymore, we should maybe only talk about the checkpoints?

File: docs/api/models/modules.rst [hunk 2]
```diff
@@ -101,31 +113,36 @@ supports CUDA Graphs and Automatic Mixed-Precision.
...
-that this model supports. In this case we will enable CUDA Graphs and Automatic Mixed-Precision.
+:class:`~physicsnemo.core.module.Module` instead of ``torch.nn.Module``. The
+:class:`~physicsnemo.core.module.Module` class acts like a direct replacement for the
+``torch.nn.Module`` and provides additional functionality for
+:ref:`saving and loading checkpoints <saving-and-loading-physicsnemo-models>`,
+:ref:`model registry <physicsnemo-model-registry>`, and more.
+Refer to the :ref:`API docs <modules-api-reference>` of
+:class:`~physicsnemo.core.module.Module` for further details.
+
+Additionally, we add a :class:`~physicsnemo.core.meta.ModelMetaData` dataclass
```
Comment (ktangsali) [inline docs/api/models/modules.rst]: Maybe let's say that this is true for the legacy models but no longer "required" if not using StaticCapture? I wonder if this reduces the barrier to entry

File: docs/api/models/modules.rst [hunk 3]
```diff
@@ -235,30 +209,88 @@ below.
...
+    PhysicsNeMoModel = physicsnemo.Module.from_torch(
+        TorchModel, meta=ConvMetaData(), name="MyConvNet"
+    )
+    PhysicsNeMoModel.__name__  # 'MyConvNet'
 
+    # Once instantiated, the result is a PhysicsNeMo Module whose class name
+    # is the one specified above.
+    model = PhysicsNeMoModel()
+
+To be able to later load this model with
```
Comment (ktangsali) [inline docs/api/models/modules.rst]: This is optional right?

File: docs/api/models/modules.rst [hunk 4]
```diff
@@ -471,11 +509,127 @@ You cannot directly pass a ``torch.nn.Module`` instance to a
     # You must use Module.from_torch() to convert it first.
 
 
+.. _backward-compatibility:
+
+Backward Compatibility
+----------------------
+
+When evolving a model class over time --- for example, renaming the class,
```
Comment (ktangsali) [inline docs/api/models/modules.rst]: nit: maybe three dashes are not needed ... ;)

File: docs/api/models/modules.rst [hunk 5]
```diff
@@ -471,11 +509,127 @@ You cannot directly pass a ``torch.nn.Module`` instance to a
...
+adding or removing constructor arguments --- you may still need to load
+checkpoints that were saved with an older version of the class. PhysicsNeMo
+provides a versioning and argument-mapping system for this purpose.
+
+The key ingredients are:
+
+- **``__model_checkpoint_version__``** --- A version string (e.g. ``"0.2.0"``)
```
Comment (ktangsali) [inline docs/api/models/modules.rst]: let's make it atleast 1.x.x and higher? Because before that, the product name was different, not sure if the checkpoints saved before that are also backwards compatible.

Comment (CharlelieLrt) [reply docs/api/models/modules.rst]: Note: this is the model version, so different from the physicsnemo version. The metadata file also contains the physicsnemo version.

IIRC the model version was introduced _after_ the product name change, and the default version for all models was `0.1.0`. I don't think we can change this default version anymore, because some checkpoints have it saved somewhere, and changing it might break them.

File: docs/api/models/modules.rst (hunk 4 above)
Comment (CharlelieLrt) [reply docs/api/models/modules.rst]: No idea what these three dashes are 😆 
Okay, removed the `---` everywhere

File: docs/api/models/modules.rst (hunk 1 above)
Comment (CharlelieLrt) [reply docs/api/models/modules.rst]: Agree!
I rephrased this to focus on checkpointing, registry, features to help support backward compatibility or model versioning.
There is still a mention of built-in performance optimization, but for the models included in the zoo only.

File: docs/api/models/modules.rst (hunk 2 above)
Comment (CharlelieLrt) [reply docs/api/models/modules.rst]: I think that's a good point.
⚠️ I removed all mention of the metadata from this file, since the metadata is only needed for static capture

File: docs/api/models/modules.rst (hunk 3 above)
Comment (CharlelieLrt) [reply docs/api/models/modules.rst]: Good catch! Yes, that's totally optional, so I watered down these explanations.

File: physicsnemo/utils/checkpoint.py [hunk 6]
```diff
@@ -361,6 +388,30 @@ def load_checkpoint(
...
+    0
+
+    Load a specific epoch:
+
+    >>> with tempfile.TemporaryDirectory() as tmpdir:
+    ...     save_checkpoint(tmpdir, models=model, epoch=0)
+    ...     save_checkpoint(tmpdir, models=model, epoch=1)
+    ...     epoch = load_checkpoint(tmpdir, models=model, epoch=0)
+    ...     epoch
+    0
```
[the remark this answers is not in the document]
Comment (CharlelieLrt) [reply physicsnemo/utils/checkpoint.py]: Good catch, I modified the example in the docstring accordingly
